symengine/lib/symengine_wrapper.pyx

```diff
@@ -1940,9 +1948,12 @@ cdef class DenseMatrixBase(MatrixBase):
     def _applyfunc(self, f):
         cdef int nr = self.nrows()
         cdef int nc = self.ncols()
+        cdef Basic e_;
         for i in range(nr):
             for j in range(nc):
-                self._set(i, j, f(self._get(i, j)))
+                e_ = _sympify(f(self._get(i, j)))
+                if e_ is not None:
+                    deref(self.thisptr).set(i, j, e_.thisptr)
```

Comment:
Above method would still allow an `ImmutableMatrix` to be mutated from python. You can copy the code from here and move it to `applyfunc` and move `_applyfunc` to `MutableDenseMatrix`.